src/core/include/iDynTree/Core/Utils.h

```diff
@@ -17,6 +17,33 @@
  */
 #define IDYNTREE_UNUSED(var) ((void)var)
 
+/**
+ * \brief Macro to deprecate functions and methods
+ *
+ * see https://blog.samat.io/2017/02/27/Deprecating-functions-and-methods-in-Cplusplus/
+ */
+// C++14
+#if __cplusplus >= 201402L
+  #if defined(__has_cpp_attribute)
+    #if __has_cpp_attribute(deprecated)
+      #define IDYNTREE_DEPRECATED [[deprecated]]
+      #define IDYNTREE_DEPRECATED_WITH_MSG(func, msg) [[deprecated(msg)]] func
```

Comment:
It's just style imo, often a message is not required, and this:
```cpp
ret_type IDYNTREE_DEPRECATED function(type1 arg1, type2 arg2);
```
is nicer than:
```cpp
ret_type IDYNTREE_DEPRECATED(function(type1 arg1, type2 arg2),"");
```